NeoML/src/TraditionalML/BytePairEncoder.cpp

```diff
@@ -153,164 +148,178 @@ void CBytePairEncoder::Decode( const CArray<int>& tokenIds,
-	const auto& inseparable = findInseparableToken( finalizedDictionary, auxTokens );
-	NeoAssert( inseparable.IsEmpty() );
-
-	tokens.SetBufferSize( finalizedDictionary.Size() );
-	// Import vocabulary
-	for( int i = 0; i < finalizedDictionary.Size(); i++ ) {
-		auto token = finalizedDictionary.GetWord( i );
-		NeoAssert( !tokenToId.Has( token ) );
 
-		// To encode input texts safely, we replace EOW/SOW with unprintable symbols.
-		if( useStartOfWordToken ) {
-			const int sowPos = token.Find( startOfWordToken );
-			if( sowPos != NotFound ) {
-				// Check that Start-of-Word isn't located in the middle of the token
-				NeoAssert( token.Find( startOfWordToken, 1 ) == NotFound );
-				token.StrReplace( 0, startOfWordToken.Length(), StartOfWordTokenInternal );
-			}
-		}
-		if( useEndOfWordToken ) {
-			const int eowPos = token.Find( endOfWordToken );
-			if( eowPos != NotFound ) {
-				// Check that End-of-Word isn't located anywhere except the end of the token
-				NeoAssert( eowPos == token.Length() - endOfWordToken.Length() );
-				token.StrReplace( eowPos, endOfWordToken.Length(), EndOfWordTokenInternal );
-			}
-		}
-
-		tokenToId.Add( token, tokens.Size() );
-		tokens.Add( token );
+	const CArray<CString> auxTokens = { params.StartOfWordToken, params.EndOfWordToken };
+	for( int i = 0; i < tokens.Size(); ++i ) {
+		NeoAssert( isValidToken( tokens[i], auxTokens ) );
 	}
 }
 
-// Returns the first inseparable token, if any
-CString CBytePairEncoder::findInseparableToken( const CWordDictionary& dictionary, const CArray<CString>& auxTokens )
+// Checks that token is a letter, auxiliary or a combination of 2 other tokens
+bool CBytePairEncoder::isValidToken( const CString& token, const CArray<CString>& auxTokens ) const
 {
-	CArray<bool> isSeparable;
-	isSeparable.Add( false, dictionary.Size() );
-
-	for( int i = 0; i < dictionary.Size(); ++i ) {
-		const auto& leftToken = dictionary.GetWord( i );
-
-		for( int j = 0; j < dictionary.Size(); ++j ) {
-			const auto& rightToken = dictionary.GetWord( j );
-			const auto& merge = MergeTokens( leftToken, rightToken );
-
-			const int mergeTokenId = dictionary.GetWordId( merge );
-			if( mergeTokenId != NotFound ) {
-				isSeparable[mergeTokenId] = true;			
-			}
-		}
+	const int charLength = UseRawBytes() ? 1 : getUtf8CharLength( token[0] );
+	if( charLength == token.Length() ) {
+		// ok, single letter
+		return true;
 	}
 
-	for( int i = 0; i < dictionary.Size(); ++i ) {
-		if( isSeparable[i] ) {
-			// ok, token is a combination of two other tokens
-			continue;
-		}
-
-		auto token = dictionary.GetWord( i );
-		const int charLength = getUtf8CharLength( token[0] );
-		if( charLength == token.Length() ) {
-			// ok, single letter
-			continue;
-		}
+	if( auxTokens.Has( token ) ) {
+		// eow/bow
+		return true;
+	}
 
-		if( !auxTokens.Has( token ) ) {
-			// not a combination, not a single letter, not auxiliary (eow/bow)
-			return token;
+	for( int j = 1; j < token.Length(); ++j ) {

```

Comment:
The iteration step which is always equal to 1 is optimal only for RawBytes mode.
Otherwise increment can be replaced with charLength (see line 261).
But in this case special auxTokens should be processed properly.